lib/internal/Magento/Framework/HTTP/Adapter/Curl.php

```diff
@@ -175,6 +175,9 @@ public function write($method, $url, $http_ver = '1.1', $headers = [], $body = '
         if ($method == \Zend_Http_Client::POST) {
             curl_setopt($this->_getResource(), CURLOPT_POST, true);
             curl_setopt($this->_getResource(), CURLOPT_POSTFIELDS, $body);
+        } elseif ($method == \Zend_Http_Client::PUT) {
+            curl_setopt($this->_getResource(), CURLOPT_CUSTOMREQUEST, 'PUT');
```

Comment:
The `CURLOPT_PUT` constant was deprecated a couple versions of cURL ago ([source](https://curl.haxx.se/libcurl/c/CURLOPT_PUT.html)), but the php.net documentation hasn't been updated yet.

Also, a lot APIs use the PUT method for other purposes than uploading files (only), for example to update an object rather than creating one. In that case `CURLOPT_POSTFIELDS` should be used. If uploading files via PUT requests does not use the `CURLOPT_POSTFIELDS` option, the function `Curl::write()` method should be modified imho, e.g. with an extra parameter.

I haven't tested this myself, but I think [`CURLFile`](http://php.net/manual/en/class.curlfile.php) can be used in combination with PUT requests too.